File: examples/typescript/middleware_and_hooks.ts
/**
 * MemoClaw middleware / hooks — logging, telemetry, request mutation.
 */
import { MemoClawClient, MemoClawError } from '@memoclaw/sdk';

const client = new MemoClawClient({ wallet: '0xYourWalletAddress' });

// ── Logging middleware ──────────────────────────────────────────────
client.onBeforeRequest((method, path, body) => {
  console.log(`→ ${method} ${path}`, body ? JSON.stringify(body).slice(0, 80) : '');
});

client.onAfterResponse((method, path, data) => {
  console.log(`← ${method} ${path} OK`);
});

client.onError((method, path, err) => {
  console.error(`✗ ${method} ${path}: [${err.code}] ${err.message}`);
});

// ── Auto-namespace injection ────────────────────────────────────────
// Mutate every store request to inject a default namespace
const namespacedClient = new MemoClawClient({ wallet: '0x...' })
  .onBeforeRequest((_method, path, body) => {
    if (path === '/v1/store' && body && typeof body === 'object') {
      return { ...body, namespace: (body as Record<string, unknown>).namespace ?? 'my-app' };
    }
  });

await namespacedClient.store({ content: 'Auto-namespaced!' });

// ── Telemetry / timing ─────────────────────────────────────────────
const timings = new Map<string, number>();

const telemetryClient = new MemoClawClient({ wallet: '0x...' })
  .onBeforeRequest((_method, path) => {
    timings.set(path, Date.now());
  })
  .onAfterResponse((_method, path) => {
    const start = timings.get(path);
    if (start) {
      console.log(`${path} took ${Date.now() - start}ms`);
      timings.delete(path);
    }
  });

await telemetryClient.recall({ query: 'test' });

// ── Batch delete ────────────────────────────────────────────────────
const results = await client.deleteBatch(['mem-1', 'mem-2', 'mem-3']);
console.log(`Deleted ${results.filter(r => r.deleted).length} memories`);

// ── search() alias (Mem0/Pinecone convention) ───────────────────────
const searchResults = await client.search({ query: 'typescript preferences' });
console.log(`Found ${searchResults.memories.length} memories via search()`);

File: examples/python/middleware_and_hooks.py
"""Using middleware hooks for logging, metrics, and request modification."""

import time
from memoclaw import MemoClaw


def log_requests(method: str, path: str, body: dict | None) -> dict | None:
    """Log every outgoing request."""
    print(f"→ {method} {path}")
    return None  # Don't modify the body


def track_latency(method: str, path: str, result):
    """Track response times (simplified — use a real metrics library in production)."""
    print(f"← {method} {path} OK")
    return None


def handle_errors(method: str, path: str, error: Exception):
    """Custom error handling — e.g., send to Sentry."""
    print(f"✗ {method} {path} failed: {error}")


# Chain hooks fluently
client = (
    MemoClaw()
    .on_before_request(log_requests)
    .on_after_response(track_latency)
    .on_error(handle_errors)
)

# Add default namespace via before_request hook
def add_default_namespace(method: str, path: str, body: dict | None) -> dict | None:
    if body and "namespace" not in body:
        body["namespace"] = "my-app"
    return body

client.on_before_request(add_default_namespace)

# All requests now go through hooks
result = client.store("Test memory with hooks", importance=0.5)
print(f"Stored: {result.id}")

client.close()
